File: 백준/Gold/18223. 민준이와 마산 그리고 건우/민준이와 마산 그리고 건우.py
import sys,heapq
input = sys.stdin.readline

# 출발지 : 1번 dijkstra 시작 지점 
# 마산 : v번 정점
# 건우: p번 정점

# 건우를 도와주는 경로 길이 <= 최단경로 길이 

n,e,p = map(int,input().split())
graph = [[] for _ in range(n+1)]
INF = int(1e9)

def dijkstra(start):
    heap = []
    distance = [INF] * (n+1)
    heapq.heappush(heap,[0,start])
    distance[start] = 0

    while heap:
        d,now = heapq.heappop(heap)
        if distance[now] < d:
            continue
        for v,w in graph[now]:
            cost = d+w
            if cost < distance[v]:
                distance[v] = cost
                heapq.heappush(heap,[cost,v])
    
    return distance

for _ in range(e):
    a,b,c = map(int,input().split())
    graph[a].append([b,c])
    graph[b].append([a,c])

# 1번 시작점 
f_dist = dijkstra(1)
# p번 시작점
p_dist = dijkstra(p)

if f_dist[p] + p_dist[n] <= f_dist[n]:
    print('SAVE HIM')
else:
    print('GOOD BYE')

File: 백준/Gold/18223. 민준이와 마산 그리고 건우/민준이와 마산 그리고 건우.js
const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
let INF = Number(1e9);

let [V,E,P] = input[0].split(' ').map(Number);
let graph = Array.from(Array(V+1),() => Array())

for(let i=1; i<E+1; i++){
    let [a,b,c] = input[i].split(' ').map(Number);
    graph[a].push([b,c]);
    graph[b].push([a,c]);
}

function dijkstra(start){
    let heap = []
    let distance = new Array(V+1).fill(INF);
    heap.push([0,start]);
    distance[start] = 0;

    while(heap.length){
        let [d,now] = heap.shift();
        if(distance[now] < d){
            continue
        }
        for(let [v,w] of graph[now]) {
            let cost = d+w;
            if(cost < distance[v]) {
                distance[v] = cost;
                heap.push([cost,v]);
            }
        }
    }
    return distance;
}

let f_dist = dijkstra(1);
let p_dist = dijkstra(P);

if (f_dist[P] + p_dist[V] <= f_dist[V]) {
    console.log('SAVE HIM');
} else {
    console.log('GOOD BYE');
}